static/js/topic_generator.js

```diff
@@ -3,31 +3,73 @@ import * as pm_conversations from "./pm_conversations";
 import * as stream_data from "./stream_data";
 import * as stream_sort from "./stream_sort";
 import * as stream_topic_history from "./stream_topic_history";
+import * as topic_list from "./topic_list";
+import * as topic_list_data from "./topic_list_data";
 import * as unread from "./unread";
 
 export function next_topic(streams, get_topics, has_unread_messages, curr_stream, curr_topic) {
+    // curr_stream_widget_id is used only for the TopicListWidget associated with left-sidebar filter
+    const curr_stream_widget_id = topic_list.active_stream_id();
     const curr_stream_index = streams.indexOf(curr_stream); // -1 if not found
 
     if (curr_stream_index >= 0) {
+        // Use 2 pass approach. First pass (filtered_topics_in_stream) iterates all topics that match
+        // the left-sidebar filter. Second pass (all_topics_in_stream) iterates all topics in the
+        // entire unfiltered stream. This 2 pass approach was made for users to jump to the next
+        // unread topic (with respect to the left-sidebar filter) in a stream using 'n' hotkey aka
+        // 'n_key', rather than jumping to the next unread topic overall, which disregards a filter.
```

Comment:
One potential gap in this approach is that probably we should respect a filter on what streams are visible, as well?  I think that's less important than the topics one, but if we can cover that as well, it'd be a fully correct implementation.